pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,180 @@
+
+        self.qinf = {k: npeval(self.cfg.getexpr(cfgsect, k), self.consts)
+                     for k in privarmap}
+        self.uinf = np.array([self.qinf[k] for k in 'uvw'[:ndims]])
+
+        if self.incomp:
+            self.qinf['rho'] = self.uinf['p']/self.consts['ac-zeta']
+            self.qinf['c'] = np.sqrt(self.consts['ac-zeta'])
+        else:
+            gamma = self.consts['gamma']
+            self.qinf['c'] = np.sqrt(gamma * self.qinf['p'] / self.qinf['rho'])
+        self.qinf['M'] = [self.qinf[k]/self.qinf['c'] for k in 'uvw'[:ndims]]
+        self.Minf = np.linalg.norm(self.qinf['M'], ord=2)
+
+        ele_map = intg.system.ele_map
+        self.emap = {k: i for i, k in enumerate(ele_map.keys())}
+        self._init_surf(ele_map)
+
+    def _init_surf(self, ele_map):
+        ndims = self.ndims
+
+        self.proj = {}
+
+        for etype, v in self._surf_region_data.items():
+            eles = ele_map[etype]
+
+            pnorm = eles._pnorm_fpts
+            rcpdjacs = eles.rcpdjac_at_np('fpts')
+
+            for fidx in set(x for i, x in v):
+                eidx = [i for i, x in v if x == fidx]
+                facefpts = eles.basis.facefpts[fidx]
+                fslice = np.ix_(facefpts, eidx)
+                npts = len(eidx)*len(facefpts)
+
+                m0 = eles.basis.m0[facefpts]
+                a = eles.basis.fpts_wts[facefpts][:, None] / rcpdjacs[fslice]
+                a = np.reshape(a, (-1,))
+                n_mag = np.linalg.norm(pnorm[fslice], ord=2, axis=-1,
+                                       keepdims=True)
+                n = np.reshape((pnorm[fslice] / n_mag).transpose(2, 0, 1),
+                               (ndims, -1))
+                dist = self._distances(etype, eles, fslice)
+
+                self.proj[etype, fidx] = (eidx, m0, n, a, dist)
+
+    def _distances(self, etype, eles, fslice):
+        spts = eles.ploc_at_np('fpts').transpose(0, 2, 1)[fslice]
+        s = np.reshape(spts, (-1, self.ndims))
+        o = self.obsv_pts
+
+        r_o = o[None, :, :] - s[:, None, :]
+        d = np.linalg.norm(r_o, ord=2, axis=-1)
+        x = r_o / d[:, :, None]
+
+        m_r = self.qinf['M'] * x
+        m_r_dot = np.sum(m_r, axis=-1)
+        m_inf = np.linalg.norm(self.qinf['M'], ord=2, axis=0)
+
+        inv_gamma = np.sqrt(1 - m_inf*m_inf)
+
+        gamma = 1 / inv_gamma
+        s_fact = np.sqrt(inv_gamma*inv_gamma + m_r_dot*m_r_dot)
+        r_star_vec = r_o * s_fact[..., None]
+        r_vec = gamma * gamma * (r_star_vec - r_o * m_r_dot[..., None])
+
+        r_star = np.linalg.norm(r_star_vec, ord=2, axis=-1)
+        r = np.linalg.norm(r_vec, ord=2, axis=-1)
+
+        r_star_vec_n = r_star_vec / r_star[..., None]
+        r_vec_n = r_vec / r[..., None]
+
+        return r_vec_n, 1/r_star, r_star_vec_n
+
+    def fwh_solve(self, intg):
+        o_vals = np.zeros(self.nobvs)
+        ci = 1 / self.qinf['c']
+
+        for (etype, fidx), p in self.proj.items():
```

Comment:
Ensure that all soln-like properties are queried outside of the loop to prevent MPI deadlocks.

Reply:
You can just do `soln = intg.soln` and `soln_t = intg.dt_soln` at the top of the function.  This ensures there will be no divergence.